Libre reports

ar4s opened the pull request:
First step to microservices based on redis and rq.

Motivation:
Separate modules like generating PDF documents from Ralph so that the module
doesn't know anything about Ralph and ralph doesn't required any extra module.
Ralph knows only name of queue where external module listening and know about
functions provided by module.

See #1791 for more information.

@mkurek @szok please review

mkurek commented on src/ralph/attachments/models.py:
```diff
@@ -41,6 +28,15 @@ def get_attachments_for_object(self, obj):
+        attachment.uploaded_by = uploaded_by
+        with open(file_path, 'rb') as f:
+            content = ContentFile(f.read())
+            attachment.file.save('test', content, save=True)
```
why test?

ar4s replied:
it's name of file it can be anything cause after save `get_file_path` override it

mkurek commented on src/ralph/reports/forms.py:
```diff
@@ -0,0 +1,12 @@
+    def clean(self):
+        defaults = [row['default'] for row in getattr(self, 'cleaned_data', [])]
+        counted = Counter(defaults)[True]
+        if counted > 1 or counted == 0:
```
maybe `if counted != 1`? :)

szok commented on src/ralph/settings/base.py:
```diff
@@ -167,3 +167,7 @@
+
+INKPY = {
+    'script_path': 'test'
+}  # REMOVE THIS
```
Remove this :)

vi4m commented:
According to the Zen of ralph, this dependency(inkpy) should be moved away from ralph base package.

szok commented on src/ralph/back_office/models.py:
```diff
@@ -185,3 +189,56 @@ def change_hostname(self, **kwargs):
+            f.write(result)
+        return add_attachment_from_disk(
+            self, output_path, request.user,
+            'Document autogenerated by {} transition.'.format(name)
```
Add translation

quamilek commented:
Please describe functionality of this PR :smile:

xliiv commented on vagrant/provision.sh:
```diff
@@ -39,3 +47,7 @@ fi
+
+# copy libreoffice service's conf
+sudo cp /home/vagrant/src/ralph/vagrant/upstart/soffice.conf /etc/init/soffice.conf
+sudo start soffice
```
If possible put all LibreOffice code into one file and call it here.

ar4s replied:
I extracted LibreOffice's stuff and moved to separated script (`vagrant/provisioning_scripts/libre_office.sh`)

szok commented on src/ralph/back_office/models.py:
```diff
@@ -185,3 +189,55 @@ def change_hostname(self, **kwargs):
+
+    def _generate_report(self, name, request):
+        report = Report.objects.get(name=name)
+        template = report.templates.first()
```
use default=True

szok commented on src/ralph/reports/forms.py:
```diff
@@ -0,0 +1,15 @@
+
+class ReportTemplateFormset(BaseInlineFormSet):
+    def clean(self):
+        defaults = [
```
check whether should be 

```
 for form in self.forms:
    form.clenaded_data
```

ar4s replied:
now it works. `self.cleaned_data` return list of `clenead_data` of each form in formset

szok commented on src/ralph/lib/external_services/base.py:
```diff
@@ -0,0 +1,49 @@
+redis_conn = Redis(**get_redis_connection_params())
+
+
+RALPH_EXTERNAL_SERVICES = getattr(settings, 'RALPH_EXTERNAL_SERVICES')
```
add default value for getattr :)

szok commented on src/ralph/lib/external_services/base.py:
```diff
@@ -0,0 +1,49 @@
+
+class ExternalService(object):
+    def __init__(self, service_name=None, name=None, method=None):
+        if not any([service_name, name, method]):
```
is needed if it is after all([[service_name, name, method]])

szok commented on src/ralph/reports/models.py:
```diff
@@ -0,0 +1,43 @@
+class Report(NamedMixin, TimeStampMixin, models.Model):
+    @property
+    def templates(self):
+        return self.templates.all() if self.templates.count() else []
```
use only .all()

vi4m commented on src/ralph/lib/external_services/base.py:
```diff
@@ -0,0 +1,49 @@
+        if not job.is_queued:
+            raise QueuedServiceError
+        while not any([job.is_finished, job.is_failed]):
+            time.sleep(.1)
```
0.1

vi4m commented on src/ralph/attachments/helpers.py:
```diff
@@ -0,0 +1,36 @@
+
+def get_file_path(instance, filename, default_dir='attachments'):
+    """
+    Function returns normalized filename with directory structure.
```
Docstring example:

def fetch_bigtable_rows(big_table, keys, other_silly_variable=None):
    """Fetches rows from a Bigtable.

```
Retrieves rows pertaining to the given keys from the Table instance
represented by big_table.  Silly things may happen if
other_silly_variable is not None.

Args:
    big_table: An open Bigtable Table instance.
    keys: A sequence of strings representing the key of each table row
        to fetch.
    other_silly_variable: Another optional variable, that has a much
        longer name than the other args, and which does nothing.

Returns:
    A dict mapping keys to the corresponding table row data
    fetched. Each row is represented as a tuple of strings. For
    example:

    {'Serak': ('Rigel VII', 'Preparer'),
     'Zim': ('Irk', 'Invader'),
     'Lrrr': ('Omicron Persei 8', 'Emperor')}

    If a key from the keys argument is missing from the dictionary,
    then that row was not found in the table.

Raises:
    IOError: An error occurred accessing the bigtable.Table object.
"""
pass
```

mkurek commented on src/ralph/lib/external_services/base.py:
```diff
@@ -0,0 +1,49 @@
+class ExternalService(object):
+    def __init__(self, service_name=None, name=None, method=None):
+        if not any([service_name, name, method]):
+            raise BadArgumentsError
```
shouldn't this exception be instantiated?

mkurek commented:
fix this minor issues, make it mergable (+ merge it into one commit) and it's :ok: for me

xliiv commented on src/ralph/settings/base.py:
```diff
@@ -172,4 +172,17 @@
+REDIS_CONNECTION = {
+    'HOST': os.environ.get('REDIS_PORT_6379_TCP_ADDR', 'localhost'),
+    'PORT': os.environ.get('REDIS_PORT_6379_TCP_PORT', '6379'),
+    'DB': 0,
```
This could also be taken from envs?

xliiv commented on src/ralph/attachments/models.py:
```diff
@@ -41,6 +28,15 @@ def get_attachments_for_object(self, obj):
+        attachment.uploaded_by = uploaded_by
+        with open(file_path, 'rb') as f:
+            content = ContentFile(f.read())
+            attachment.file.save('', content, save=True)
```
`''`?

xliiv commented on src/ralph/reports/models.py:
```diff
@@ -0,0 +1,43 @@
+class Report(NamedMixin, TimeStampMixin, models.Model):
+    @property
+    def templates(self):
+        return self.templates.all()
```
why this?

xliiv commented on src/ralph/reports/models.py:
```diff
@@ -0,0 +1,43 @@
+
+class Report(NamedMixin, TimeStampMixin, models.Model):
+    @property
+    def templates(self):
```
why this?

xliiv commented on src/ralph/lib/external_services/base.py:
```diff
@@ -0,0 +1,49 @@
+
+class ExternalService(object):
+    def __init__(self, service_name=None, name=None, method=None):
+        if not any([service_name, name, method]):
```
I dont get this concept of `name`, maybe docstring would fix me ;)

ghost commented:
:ok:

Commit:
added mechanism to generate report by microservice

    First step to microservices based on redis and rq.

    Motivation:
    Separate modules like generating PDF documents from Ralph so that the module
    doesn't know anything about Ralph and ralph doesn't required any extra module.
    Ralph knows only name of queue where external module listening and know about
    functions provided by module.

    See #1791 for more information.
